### ISSUE TITLE
conflict between a hyphen (-) in the version and rpmbuild

### ISSUE THREAD

**Issue by markgalassi** [semver/semver#145]
I decided to use your semver approach in its full glory for a new significant project.  I think it's an important thing, and it would be good for people to just adopt it so we don't have so much variability.

I have run in to one problem: I tagged a 0.1.0-alpha.0 release and I also set that to be the version in configure.ac (I'm using automake/autoconf) like this:

AC_INIT([myproject], [0.1.0-alpha.1], [markgalassi@lanl.gov])

unfortunately this brings out an error when I build my tarball myproject-0.1.0-alpha.1.tar.gz and then build an RPM with

rpmbuild -ta myproject-0.1.0-alpha.1.tar.gz

I get this error:

[markgalassi@beachhead _build]$ rpmbuild -ta myproject-0.1.0-alpha.1.tar.gz 
error: line 3: Illegal char '-' in: Version: 0.1.0-alpha.1
[markgalassi@beachhead _build]$ 

I build myproject.spec from a myproject.spec.in file which starts with:

Name: @PACKAGE@
Version: @VERSION@
%define dateNoDashes %(date +"%Y%m%d")
Release: %{dateNoDashes}%{?dist}

and the myproject.spec comes out to be:

Name: myproject
Version: 0.1.0-alpha.1
%define dateNoDashes %(date +"%Y%m%d")
Release: %{dateNoDashes}%{?dist}

This seems to point to a conflict between a rather natural way of doing RPM specs and semver.  I wonder how other people have resolved this.

Right now I'm using 0.1.0alpha.1 as the version, but I'd rather stick to semver's spec.

Another option would be to remove the hyphen when generating the spec file.

**Comment by vincentbernat** [comment]
TL;DR: don't use `~` blindly. Use `~` for alpha/beta/rc. Use `+` for snapshot.

I stumbled randomly on this issue and while it is a bit old, maybe I could share some context. As Debian (and I would like to believe this is the same for other systems) packagers, we don't really care about semantic versioning. This is something that should be solved upstream. We only care that version numbers increase with each release.

With RPM format, the only forbidden character is `-` because it is the separator with the distribution version (which permits to record packaging changes). With DEB format, the version number can only contain alphanumerics and `.`, `+`, `-`, `:`, `~`. So, it is more restricted but the hyphen is allowed. This is because it is not allowed in the distribution version. So, there is a divergence between RPM and DEB. Both of them use the hyphen as separator but RPM considers the first hyphen to be the separation between the upstream version and the distribution version and DEB considers this is the last hyphen. Hence, when we have `0.6.3-alpha1.14-1`, for an RPM based system, upstream is `0.6.3` and distribution is `alpha1.14-1` while on a Debian-based system, it is `0.6.3-alpha1.14` and `1`

All tools are using the existing convention, so it is absolutely not possible to change it.

However, this was never a problem. Let's say that the upstream version is `0.6.3-alpha1.14`. We just transform it manually to `0.6.3~alpha1.14`. Problem solved for us. We only have a few handful tools manipulating directly the upstream version, so this is not a problem (one of those tools is `uscan` which tells if there is a newer version available. For that, it needs to mangle the upstream version with a regex).

As I said before, the most important point for us is for the versions to be correctly sorted. So, replacing hyphens by another sign should not be done blindly. For alpha/beta/rc, we use `~` because it orders before any other characters even the "no" character. So, `0.6.14~alpha1` < `0.6.14~alpha2` < `0.6.14~beta1` < `0.6.14`. On the other hand, when the version is a snapshot, we must ensure that the order is also correct, so it is `0.6.14` < `0.6.14+git20150604+5` (assuming there is 5 commits since 0.6.14) < `0.6.14+git20140504+6`. We don't really have a convention for this. The only point is to ensure that versions are always increasing.

If we make an error, we can reset everything by using an epoch (this is specific to DEB, I think there is no such thing for RPM). Suppose we have released `0.6.14-alpha1-1` and `0.6.14` is released. We cannot package it as `0.6.14-1` (which is inferior), so we use `1:0.6.14-1`. But this is really a last resort solution because if another package requires `< 0.6.15`,  `1:0.6.14-1` wouldn't match.

**Comment by Wuzzy2** [comment]
Maybe we should consider RPM package version numbers to be more or less independent of the actual software versions; the packages use their own versioning system but packagers seem to try hard to come as close to the original version number as possible.
If we view it this way, then it is no problem that RPM package versions may not perfectly map to the actual software version.
I suppose SemVer-based software is by far not the only software of which its version number can't be perfectly translated into a RPM version number. The hyphen is a common character in software versions anyways.

**Comment by lukeyeager** [comment]
> Apart from that, changing Semver in the suggested manner would be another breaking change. People would start to ask why a standard, which is aimed at settling a certain issue once and for all, keeps on introducing backwards-incompatible changes, thus not settling anything at all.

I'm with @tbull - keep SemVer pure and strict. Don't change it.

> Maybe we should consider RPM package version numbers to be more or less independent of the actual software versions; the packages use their own versioning system but packagers seem to try hard to come as close to the original version number as possible.
> If we view it this way, then it is no problem that RPM package versions may not perfectly map to the actual software version.

I'm with @Wuzzy2 - follow SemVer in your project, and follow it as closely as possible when dealing with package mgmt.

=====

To add some more information to this discussion, [debian packages](https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-Version) have no problem with a version like `0.14.0-beta-1`. `upstream_version` is correctly parsed as `0.14.0-beta` and `debian_revision` is correctly parsed as `1`. [The problem](https://lintian.debian.org/tags/rc-version-greater-than-expected-version.html) is when you try to use a version like `0.14.0-beta.1-1`:

    rc-version-greater-than-expected-version 0.14.0-beta.1 > 0.14.0 (consider using 0.14.0~beta.1)

The versions are still *parsed* correctly, but they are *sorted* incorrectly. Lintian helpfully lets you know that this is going to be a problem.

Switching the version to `0.14.0~beta.1-1` for the debian package and suppressing the [`package-name-doesnt-match-sonames`](https://lintian.debian.org/tags/package-name-doesnt-match-sonames.html) warning solves my problem. I'm not overly worried about the debian package version for a pre-release not matching my project's semantic version exactly.

**Comment by jasonmp85** [comment]
I came across this thread while researching the subject and wanted to drop my proposals here.

I'm going to number the project in question in a way compliant with semver. Everywhere possible (git tags, certain packaging systems, the version reported by the software itself, etc.), I'll use the compliant semver version formulation.

But in the `.deb` and `.rpm` I have to make concessions to those systems. In particular, it's easy enough for a `.deb` to have version `1.0.0`, but for `2.0.0-rc.1`, I'll use `2.0.0~rc.1`. This mapping is bijective.

Similarly, in `rpm`, I can't have _any_ separator other than a period. Again, perfectly fine for stable releases, but for prerelease versions, [they say to use the release field](https://fedoraproject.org/wiki/Packaging:NamingGuidelines#Pre-Release_packages). So `2.0.0-rc.1` will instead become `VERSION=2.0.0` and `RELEASE=0.1.rc.1` (in particular, I plan on doing `RELEASE=0.`_n_`.rc.`_n_ for release candidate _n_).

Since I generally only have stable releases and release candidates (names like "alpha", "beta", etc. are avoided), this solution works fine. I'll have to think about nightly/snapshot releases at some point in the future, but thought I could post my solution here in case any else hits here via Google.

**Comment by majgis** [comment]
Use of the RELEASE field as @jasonmp85 describes above for RPMs seems like it might be on it's way to being "correct", but I needed a pragmatic alternative since we already have an internal standard for the RELEASE field.

Using the following progression from semver.org:
 
1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0

I created RPM's for equivalent versions of each of these, replacing - with ~.  The only one that did not install (trump a previous install) was 1.0.0-alpha.beta.

Therefore, from a pragmatic perspective, replacing the dash with the tilde works well, but it does not cover SemVer cases where a secondary alpha/beta is being used.

**Comment by jwdonahue** [comment]
@markgalassi 

My two cents: Implementation specific mappings between SemVer and other schemes are not strictly ruled out by the SemVer spec, provided the SemVer version of the version string can be readily acquired, either by lookup or conversion.  I have used this approach to create path names to "package" directories in the past and I can't see any reason why it would work with inflexible package managers as well.

It seems the spec was not altered and there is little chance of your requested changes being accepted. Unless you intend to provide PR for this issue, please close it at your earliest possible convenience.

**Comment by silkentrance** [comment]
@markgalassi autoconf is what you want to renovate. not the other way around. otherwise, indirection is your friend.

@jwdonahue please close this regardless.

**Comment by david-a-wheeler** [comment]
The system package managers and build sysyems are unlikely to be changed.  A simple solution is to allow tilde (\~) or underscore (\_) as a separator in SemVer for this kind of information.

**Comment by grv87** [comment]
Some devops tools allow usage of `-` in the names or versions but prohibit `+`. For example, Amazon AMIs, VirtualBox VMs or Vagrant boxes. So I substitute `_` instead of `+` to denote build metadata (namely, architecture e.g. `amd64`).

I wish there would be no overlapping changes in standard so that `_` would be used for prerelease. I think deviations in delimiters should be left for particular implementations of semver and not be a part of the standard.
Maybe to avoid such issues the standard could use abstract `prerelease delimiter` and `build metadata delimiter` terms and include section about reference/default implementation instead which could specify old default/recommended delimiters.

**Comment by jwdonahue** [comment]
@markgalassi, can you please close this issue at your earliest possible convenience.  I don't have the sauce and we're trying to reduce number of none actionable issues.  Thank you.

**Comment by markgalassi** [comment]
David> One solution, for example, is to allow tilde instead of dash
    David> for this case. I imagine there are other solutions.

That's what I had proposed.  I recommend it in my article on release
engineering.  Works well with RPM but it seems to be a bit annoying for
debian.

**Comment by alexandrtovmach** [comment]
Thank you guys for continuing discussion, for now I'm only remove `staled` label, but a bit later will join your chat 👍

**Comment by jbigot** [comment]
From my point of view, there is no way to have a single standard to rule them all that would work everywhere. The approach that would seem correct to me would be to:
1. define an abstract version scheme (the various elements of a version, the ordering, etc.)
2. support multiple concrete version syntax, with a bijective system to map from the concrete version string to the abstract version value,
3. define a default concrete syntax that would be the current version syntax

Then, it would be possible to add additional concrete syntaxes for systems that do not accept the default one, for `.deb`, it would for example be a simple question of using a tilda istead of a dash (an underscore would result in incorrect ordering).

**Comment by david-a-wheeler** [comment]
@jbigot - it would have been possible, but that's neither here nor there.

Defining a standardized mapping would be a good thing, that would probably resolve it.

**Comment by Conan-Kudo** [comment]
RPM's versioning rules are as follows

* RPM splits versions in to sections which are compared piece by piece
* Version/Release strings allowed to have ASCII letters (`a-zA-Z`), digits (`0-9`) and separators (`._+~^`)
* Tilde (`~`) and Caret (`^`) are special separators and have special meaning/handling
* Separators are ignored by RPM's comparison - `1+0 == 1.0`
* Letters are sorted before digits - `a < 1`
* Letters are compared by their ASCII value - `a (97) > A (65)`
* Additional character makes version being higher - `1.1 < 1.10`, `1.0 < 1.0.0`
* Tilde (`~`) makes version being lower - `1.0 > 1.0~`, `1.0 > 1.0~rc1`
* Caret (`^`) makes version being higher - `1.0 < 1.0^`, `1.0 < 1.0^post1`
* Zero (`0`) after separator handled natively - `1.0 == 1.000`

**Comment by alexandrtovmach** [comment]
>SemVer spec is something stable and strict, and if you want to change it, please create fork or start another project

closing in favor of https://github.com/semver/semver/issues/703#issuecomment-841420754

**Comment by gh-andre** [comment]
Interesting thread. Long one too. A few comments above describe that translation is needed between semver and package versioning, but stop short of saying (unless I missed any) that while semver and package versioning intersect in some aspects, they actually represent two different concepts and there's no selection of punctuation that can "fix" that in a way that semver can be used verbatim in package names.

Semantic Versioning describes the level of change in the upstream software and works outside of any package versioning. That's what the upstream software may print when `--version` is used. Package managers like RPM are meant for packaging upstream software and need to accommodate a large spectrum of upstream versioning schemes to sort package versions properly within their very diverse ecosystems.

Package maintainers are often different from upstream software maintainers and may need to repackage _same version_ of upstream software, which may happen without any code changes (e.g. use different compiler flags) or because they introduced a patch on top of the patch level in the upstream software. That patch may be submitted back to upstream maintainers and may or may not make it into the software patch level. Both are patch levels, but they are not quite the same kind of a _patch_ conceptually.

Semver works great for software versioning, but when used as a package version, will limit packages to contain only changes from the upstream software maintainers. NPM is like that - all changes must come from software maintainers because there is no way to accommodate repackaging, snapshots and 3rd-party patches. It works out for NPM, but not quite so for Nuget, which has many packages created by 3rd-party package maintainers, but offers no way to repackage same software because it lacks package revisions.

Semantic versions simply need to be translated for packaging and if a build pipeline produced an artifact `myapp-1.2.0-beta1+045.tar.gz`, it may be packaged as `myapp-1.2.0-0.1.beta1.el8.rpm`, may be repackaged without any upstream source changes as `myapp-1.2.0-0.2.beta1.el8.rpm` and, when upstream changes, packaged as `myapp-1.2.0-0.3.beta2.el8.rpm`. None of these package version variations takes away from the semantic meaning of the enclosed software version.
